File: next-tauri/src/platform/contracts/vault.ts
export interface VaultFileEntry {
  relativePath: string
  sizeBytes: number
  modifiedAt: number
}

export interface VaultDocument {
  relativePath: string
  content: string
  fingerprint: string
  modifiedAt: number
}

export interface VaultSaveRequest {
  relativePath: string
  content: string
  expectedFingerprint: string | null
}

export interface VaultTrashResult {
  relativePath: string
  recoveryPath: string
}

export interface VaultGitStatus {
  available: boolean
  repository: boolean
  branch: string
  dirty: boolean
  changedFiles: number
  ahead: number
  behind: number
  remote: string
}

export interface VaultGitCommit {
  hash: string
  author: string
  timestamp: number
  subject: string
}

export interface VaultGitDetails {
  diff: string
  commits: VaultGitCommit[]
}

export interface VaultSnapshot {
  rootPath: string | null
  files: VaultFileEntry[]
  directories: string[]
  git: VaultGitStatus
}

export type VaultGitOperation = 'init' | 'pull' | 'commit' | 'push'

export interface VaultGitRequest {
  requestId: string
  operation: VaultGitOperation
  message?: string
  editorDirty: boolean
}

export interface VaultGitResult {
  operation: VaultGitOperation
  success: boolean
  exitCode: number | null
  stdout: string
  stderr: string
}

export interface VaultChangedEvent {
  reason: string
}

export interface VaultApi {
  chooseRootDirectory(): Promise<string | null>
  snapshot(): Promise<VaultSnapshot>
  configure(rootDirectory: string): Promise<VaultSnapshot>
  disconnect(): Promise<void>
  read(relativePath: string): Promise<VaultDocument>
  save(request: VaultSaveRequest): Promise<VaultDocument>
  createDirectory(relativePath: string): Promise<string>
  move(relativePath: string, destinationPath: string, expectedFingerprint: string | null): Promise<string>
  duplicate(relativePath: string, destinationPath: string, expectedFingerprint: string): Promise<VaultDocument>
  delete(relativePath: string, expectedFingerprint: string): Promise<VaultTrashResult>
  deleteEntry(relativePath: string, expectedFingerprint: string | null): Promise<VaultTrashResult>
  gitStatus(): Promise<VaultGitStatus>
  gitDetails(relativePath?: string): Promise<VaultGitDetails>
  configureGitRemote(remote: string): Promise<VaultGitStatus>
  runGit(request: VaultGitRequest): Promise<VaultGitResult>
  cancelGit(requestId: string): Promise<boolean>
  subscribe(listener: (event: VaultChangedEvent) => void): Promise<() => void>
}

File: next-tauri/src-tauri/src/contracts/vault.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultFileEntry {
    pub relative_path: String,
    pub size_bytes: u64,
    pub modified_at: u64,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultDocument {
    pub relative_path: String,
    pub content: String,
    pub fingerprint: String,
    pub modified_at: u64,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultSaveRequest {
    pub relative_path: String,
    pub content: String,
    pub expected_fingerprint: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultTrashResult {
    pub relative_path: String,
    pub recovery_path: String,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultSnapshot {
    pub root_path: Option<String>,
    pub files: Vec<VaultFileEntry>,
    pub directories: Vec<String>,
    pub git: VaultGitStatus,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultGitStatus {
    pub available: bool,
    pub repository: bool,
    pub branch: String,
    pub dirty: bool,
    pub changed_files: usize,
    pub ahead: u32,
    pub behind: u32,
    pub remote: String,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultGitCommit {
    pub hash: String,
    pub author: String,
    pub timestamp: i64,
    pub subject: String,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultGitDetails {
    pub diff: String,
    pub commits: Vec<VaultGitCommit>,
}

#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum VaultGitOperation {
    Init,
    Pull,
    Commit,
    Push,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultGitRequest {
    pub request_id: String,
    pub operation: VaultGitOperation,
    pub message: Option<String>,
    pub editor_dirty: bool,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultGitResult {
    pub operation: VaultGitOperation,
    pub success: bool,
    pub exit_code: Option<i32>,
    pub stdout: String,
    pub stderr: String,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultChangedEvent {
    pub reason: String,
}
